### ISSUE TITLE
Can't access data inside relationship fields when in a preview session

### ISSUE THREAD

**Issue by mrseanbaines** [prismicio/gatsby#422]
I'm facing an issue where I can't access data inside relationship fields when in a preview session.

I am querying a blog post which has a relationship field for author, category etc. I'm using a GraphQL fragment (doesn't matter if it's inline or not) inside of the `document` field to access the author type and query fields inside of that, e.g. `name`. This works fine outside of a preview session, but when previewing, the `document` property isn't there (it's `undefined` if I try to access it). I seem to be getting a Proxy object under the author query, but this only seems to have the built-in properties and nothing from my document type that I'm trying to query (e.g. `name`).

I'm using the `_previewable` field both on the blog post _and_ the author documents. I'm also querying for `id` everywhere possible as I think I read somewhere that this is necessary(?). I'm also using the `withPrismicPreview` HOC in my template.

This is the query exactly as I am doing it in the code:
<img width="1150" alt="image" src="https://user-images.githubusercontent.com/24367010/128684960-0182fc3b-d5c9-49db-b8df-24a7db8dfeaa.png">

This is the response outside of a preview session (same as in GraphiQL editor):
<img width="815" alt="image" src="https://user-images.githubusercontent.com/24367010/128684984-827510bb-d0e9-4827-b9a8-6f7e5d493f22.png">

This is the response when inside a preview session (this also seems to return everything from the document, rather than just what I queried for):
<img width="941" alt="image" src="https://user-images.githubusercontent.com/24367010/128685051-d2af007d-c17f-4084-8856-677d01db0f26.png">

### Versions

- gatsby-source-prismic: v4.0.1
- gatsby-plugin-prismic-previews: v4.0.1
- node: v14.17.1

### Steps to reproduce
Set up Prismic previews, query a relationship field, start a preview session, try to access the data for the linked document.

### What is expected?
I should be able to access data for relationship fields inside Primsic preview the same way I can outside of it

### What is actually happening?
I can't access data for relationship fields inside Primsic preview at all

**Comment by mrseanbaines** [comment]
**Update:**

If I log out `data.prismicBlogPost.data.author.document?.data.name` _(note the optional chaining on `document`)_, then it seems to work in the end. It looks like at some point in the preview generation lifecyle it becomes `undefined` before being defined again, so if you just try to access `document` directly then it throws and the preview doesn't work at all.

<img width="86" alt="image" src="https://user-images.githubusercontent.com/24367010/128690233-85e4dc8c-9c7c-466a-a404-a296b34f8484.png">

**Comment by angeloashmore** [comment]
@mrseanbaines Thanks for the update. That's interesting that it shows as undefined for a single render. I will take a closer look at why this is happening for you.

In the meantime, can you test this with the latest version of the plugin (currently `v4.0.3`)? This is just to rule out any recent fixes as being the cause.

---

For some background info on the plugin:

The plugin is designed to match your static query during a preview as closely as possible. For `document` fields, this is done by using a Proxy to link the `document` field to a Prismic document in the plugin's cache. When you log a content relationship field, you will see "Proxy" and won't see the `document` field. It will work at runtime since the Proxy sees the call to `document` and returns the correct value.

During previews, the whole document will be present as a performance optimization. It won't match your query exactly (which only contains a subset of the document), but your templates will still act the same since it will only be accessing the same subset of fields.

**Comment by mrseanbaines** [comment]
Thanks @angeloashmore - will try out the latest version tomorrow and get back to you. Turned out not to be a huge issue after I worked out that it eventually populates the content, so just needed a few optional chainings and/or some early returns

**Comment by angeloashmore** [comment]
Good to hear! In general, I would recommend always putting a null check around fields (including content relationship fields) since editors in Prismic can always save a document without filling in the field. This doesn't fix the root issue you describe, but it helps cover it up.

**Comment by OleksandrRakovets** [comment]
Hi @angeloashmore 

I am experiencing the same issue on the latest version of `gatsby-plugin-prismic-previews": "4.1.2"`

<img width="642" alt="Screenshot 2021-08-29 at 19 08 58" src="https://user-images.githubusercontent.com/5619390/131257348-313635cf-1a80-4acc-8e64-dbd85bfc135d.png">

Output is from `console.log(data.author.document)`. To make matters worse, `undefined`s come last now, so optional chaining has no effect.

The query works fine without preview.
<img width="898" alt="Screenshot 2021-08-29 at 19 14 32" src="https://user-images.githubusercontent.com/5619390/131257482-df3ed3d5-6b0f-485f-8951-861df1e906f8.png">

Should I omit output of relationship components entirely in preview sessions for the time being?

**Comment by angeloashmore** [comment]
Hi @OleksandrRakovets, sorry you're seeing this error during previews. This should not be happening, of course.

I would recommend doing a boolean check for content relationship fields. This will at least stop your site from crashing for now and will help in the future if a content writer does not provide a value for that field.

I will look into this more to figure out what is happening.

@OleksandrRakovets It might be easiest to figure out the issue if I can look at your code and Prismic repository. Can you send me a private message on the Prismic community forum?

To send me a message: https://community.prismic.io/new-message?username=angeloashmore&title=GitHub%20Issue%20422

**Comment by angeloashmore** [comment]
@OleksandrRakovets I sent you a message on the Prismic community forum, but I'll repost the info here for others.

> Could you try out the latest version of the Gatsby plugins?
>
> - `gatsby-source-prismic`: `v4.1.2` (already installed)
> - `gatsby-plugin-prismic-previews`: `v4.1.3`
> 
> Indeed there was a bug where the preview data was being merged into the static data before all the preview data was available. If you are curious, this is the fix: https://github.com/prismicio/prismic-gatsby/commit/4b9bfe4594810bf3b4de919ed0b3c967c20b4382

I'll close this issue as the fix *should* also fix the original issue. Basically, if the rendered component does not throw an error, the preview would eventually resolve to the correct data. With the latest version, it now will only add preview once everything is ready.

**Comment by riccoski** [comment]
I'm having this issue on `"gatsby-plugin-prismic-previews": "^5.2.1",`
